lib/JSON/Validator.pm

```diff
@@ -574,69 +573,101 @@ sub _validate {
 
 sub _validate_all_of {
   my ($self, $data, $path, $rules) = @_;
-  my $type = data_type $data, $rules;
-  my (@errors, @expected);
+  my (@errors);
 
   my $i = 0;
   for my $rule (@$rules) {
     next unless my @e = $self->_validate($_[1], $path, $rule);
-    my $schema_type = schema_type $rule;
-    push @expected, $schema_type if $schema_type;
-    push @errors, [$i, @e] if !$schema_type or $schema_type eq $type;
+    push @errors, [$i, @e];
   }
   continue {
     $i++;
   }
 
-  return E $path, [allOf => type => join('/', uniq @expected), $type]
-    if !@errors and @expected;
-  return prefix_errors allOf => @errors if @errors;
-  return;
+  return if not @errors;
+
+  # combine all 'type' errors together
+  return E $path,
+    [
+    allOf => type => join('/', uniq map $_->[1]->details->[0], @errors),
+    $errors[-1][1]->details->[2]
+    ]
+    if @errors > 1 and not grep $_->[1]->details->[1] ne 'type', @errors;
+
+  return prefix_errors allOf => @errors;
+}
+
+sub _validate_any_of_types {
+  my ($self, $data, $path, $rules) = @_;
+  my @errors;
+
+  for my $rule (@$rules) {
+    my @e = $self->_validate($_[1], $path, $rule);
+    return unless @e;
+    push @errors, @e;
+  }
+
+  if (my @e = grep $_->details->[1] ne 'type', @errors) {
+    return @e;
```

Comment:
I think this is not what you want. Why filter out the "type" errors? See my comment below for a fixed version.

Reply:
Filtering out the type errors resolves the complaint you had about the 'enum' tests, where we were reporting a mismatched type even though the other choice of type did match. (t/jv-anyof.t line 70)